### PR TITLE
add index type to the array type

### PR DESCRIPTION
**Pull request by kroening**
This commit addresses a gap between the arrays of the C programming language and standard array theories.  In C, array indices are subject to arithmetic promotion, and thus, may have a variety of types, depending on the operand type.  Standard array theories expect one specific index type per array type.

The C front-end now annotates the array type with that index type.  We will eventually enforce that all index expressions use this type.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/ansi-c/c_typecheck_type.cpp [hunk 1]
```diff
@@ -560,10 +557,15 @@ void c_typecheck_baset::typecheck_array_type(array_typet &type)
     throw 0;
   }
 
-  // check size, if any
+  // We add the index type.  This is added as a comment now for backwards
+  // compatibility, but will eventually be the first subtype.
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_type.cpp]: I'm worried that this comment will eventually (hopefully) be inconsistent with reality and no one will notice. I think this comment should only go in the comments of `array_typet`, but not in any place that (correctly) uses the API.

File: src/solvers/smt2/smt2_conv.cpp [hunk 2]
```diff
@@ -5127,8 +5127,9 @@ void smt2_convt::convert_type(const typet &type)
     // we always use array theory for top-level arrays
     const typet &subtype = array_type.element_type();
 
+    // Arrays map the index type to the element type.
     out << "(Array ";
-    convert_type(array_type.size().type());
+    convert_type(array_type.index_type());
```
Comment (tautschnig) [inline src/solvers/smt2/smt2_conv.cpp:5132]: Wasn't `.size().type()` correct here?

**Review by tautschnig — APPROVED** [review]

**Review by peterschrammel — APPROVED** [review]

File: src/solvers/smt2/smt2_conv.cpp (hunk 2 above)
Comment (kroening) [reply src/solvers/smt2/smt2_conv.cpp:5132]: It only works when the array has a size.

File: src/ansi-c/c_typecheck_type.cpp (hunk 1 above)
Comment (kroening) [reply src/ansi-c/c_typecheck_type.cpp]: Ok, will move.

File: src/solvers/smt2/smt2_conv.cpp (hunk 2 above)
Comment (tautschnig) [reply src/solvers/smt2/smt2_conv.cpp:5132]: What information does SMT-LIB require at this point? Agreed that not all arrays have a size, which may be a problem, but I'm just not sufficiently knowledgable in SMT-LIB to tell whether this change really makes sense. If that's just a workaround for a missing size then I remain unconvinced this is the right change to make.

Comment (kroening) [reply src/solvers/smt2/smt2_conv.cpp:5132]: SMT-LIB does require the index type; arrays are maps from indices to the codomain type. Using the type of the size was a workaround.

**Pull request merged**